File: leetcode/java/0543. Diameter of Binary Tree.java
package leetcode.java;

// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public int diameterOfBinaryTree(TreeNode root) {
        int[] maxD = new int[1];
        this.helper(maxD, root);
        return maxD[0];
    }

    public int getHeight(TreeNode root) {
        if (root == null) {
            return 0;
        }
        return 1 + Math.max(this.getHeight(root.left), this.getHeight(root.right));
    }

    public void helper(int[] maxD, TreeNode source) {

        if (source == null) {
            return;
        }
        maxD[0] = Integer.max(getHeight(source.left) + getHeight(source.right), maxD[0]);
        this.helper(maxD, source.left);
        this.helper(maxD, source.right);
    }
}

File: leetcode/python/0543. Diameter of Binary Tree.py
'''
count the number of edges not just from the root, but from the subtrees
'''

class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    def diameterOfBinaryTree(self, root):
        res = [0]
        self.countEdges(root, res)
        return res[0]
        
    def countEdges(self, root, res):
        if root == None:
            return 0
        l = self.countEdges(root.left, res)
        r = self.countEdges(root.right, res)
        res[0] = max(res[0], l+r)
        return 1 + max(l, r)